[FLINK-20347] Rework YARN documentation page (#14238)
apache/flink: docs/ops/deployment/yarn.md

- Make sure your YARN cluster is ready for accepting Flink applications by running `yarn top`. It should show no error messages.
- Download a recent Flink distribution from the [download page]({{ site.download_url }}) and unpack it.
- **Important** Make sure that the `HADOOP_CLASSPATH` environment variable is set up (for example by running `echo $HADOOP_CLASSPATH`). If not, set it up using 

{% highlight bash %}
export HADOOP_CLASSPATH=`hadoop classpath`
{% endhighlight %}

### Starting a Flink Session on YARN

Once you've made sure that the `HADOOP_CLASSPATH` environment variable is set, we can launch a Flink on YARN session, and submit an example job:

{% highlight bash %}

# we assume to be in the root directory of the unzipped Flink distribution

> 

Author: Yes, this is very good feedback. I was thinking about this a lot.

### Resource Allocation Behavior

A JobManager submitted in YARN session mode will wait for job submissions before starting to request and allocate TaskManagers. The other deployment modes are immediately requesting the required TaskManagers from YARN.

The memory configurations for JobManager and TaskManager processes will be respected by the YARN implementation. The number of reported VCores is by default equal to the number of configured slots per TaskManager. The [`yarn.containers.vcores`]({% link ops/config.md %}#yarn-containers-vcores) allows overwriting the number of vcores with a custom value. In order for this parameter to work you should enable CPU scheduling in your cluster.

Failed containers (including the JobManager) are replaced by YARN. The maximum number of JobManager container restarts is configured via `yarn.application-attempts` (default 1). The YARN Application will fail once all attempts are exhausted.

### High-Availability on YARN

> For enabling HA on YARN, the `yarn.application-attempts` need to be greater 1, and a [high availability service needs to be configured]().

Author: I will fill in the link now.

### Supported Hadoop versions.

Flink on YARN is compiled against Hadoop 2.4.1, which is also the earliest version we are supporting. 

For providing Flink with the required Hadoop dependencies, we recommend setting the `HADOOP_CLASSPATH` environment variable. 

> If that is not possible, the dependencies can also be put into the `lib/` folder of Flink. Flink also offers pre-bundled Hadoop fat jars for placing them in the `lib/` folder, in the "Downloads / Additional Components" section of the website. These pre-bundled fat jars are shaded to avoid dependency conflicts with common libraries.
> 
> The Flink community is not testing the YARN integration against these pre-bundled jars. 

Reviewer: ```suggestion
If that is not possible, the dependencies can also be put into the `lib/` folder of Flink. 

Flink also offers pre-bundled Hadoop fat jars for placing them in the `lib/` folder, in the "Downloads / Additional Components" section of the website. These pre-bundled fat jars are shaded to avoid dependency conflicts with common libraries. The Flink community is not testing the YARN integration against these pre-bundled jars. 
```

I would organize the paragraphs differently: The first paragraph is about integrating Hadoop. The second paragraph is about using pre-bundled jars.

Additionally, is there a reason why we don't link the Downloads page?

### Running Flink on YARN behind Firewalls

Some YARN clusters use firewalls for controlling the network traffic between the cluster and the rest of the network.
In those setups, Flink jobs can only be submitted to a YARN session from within the cluster's network (behind the firewall).
If this is not feasible for production use, Flink allows to configure a port range for its REST endpoint, used for the client-cluster communication. With this range configured, users can also submit jobs to Flink crossing the firewall.

> The configuration parameter for specifying the REST endpoint port is the following:
> 
>  * `rest.bind-port`
> 
> This configuration option accepts single ports (for example: "50010"), ranges ("50000-50025"), or a combination of
> both ("50010,50011,50020-50025,50050-50075").

Reviewer: ```suggestion
The configuration parameter for specifying the REST endpoint port [rest.bind-port]({% link ops/config.md %}#rest-bind-port). This configuration option accepts single ports (for example: "50010"), ranges ("50000-50025"), or a combination of
both ("50010,50011,50020-50025,50050-50075").
```
IMHO, we don't need to use a list if there's only one value.

> Please make sure that the configuration option `rest.port` has not been specified, because it has precedence over `rest.bind-port` and accepts no ranges.

Reviewer: ```suggestion
Please make sure that the configuration option [rest.port]({% link ops/config.md %}#rest-port) has not been specified, because it has precedence over `rest.bind-port` and accepts no ranges.
```

I added the link for `rest.port` here as well.

### User jars & Classpath

> By default Flink will include the user jars into the system classpath when running a single job. This behavior can be controlled with the `yarn.per-job-cluster.include-user-jar` parameter.

Reviewer: ```suggestion
By default Flink will include the user jars into the system classpath when running a single job. This behavior can be controlled with the [yarn.per-job-cluster.include-user-jar]({% link ops/config.md %}#yarn-per-job-cluster-include-user-jar) parameter.
```

When setting this to `DISABLED` Flink will include the jar in the user classpath instead.

> The user-jars position in the class path can be controlled by setting the parameter to one of the following:
> 
> - `ORDER`: (default) Adds the jar to the system class path based on the lexicographic order.
> - `FIRST`: Adds the jar to the beginning of the system class path.
> - `LAST`: Adds the jar to the end of the system class path.

Reviewer: ```suggestion
The user-jars position in the classpath can be controlled by setting the parameter to one of the following:

- `ORDER`: (default) Adds the jar to the system classpath based on the lexicographic order.
- `FIRST`: Adds the jar to the beginning of the system classpath.
- `LAST`: Adds the jar to the end of the system classpath.
```
Changed "class path" to "classpath".
